### PR TITLE
Generate XML test results from unittests and upload test results to CircleCI

### PR DESCRIPTION
**Pull request by dschuff**

### CODE REVIEW

**Comment by dschuff** [comment]
This allows Circle to provide summaries such as https://app.circleci.com/pipelines/github/emscripten-core/emscripten?branch=xmlrunner and https://app.circleci.com/pipelines/github/emscripten-core/emscripten/24895/workflows/9ff01283-5027-4a12-a2ff-a2c7ea30edc5/jobs/587697 and analysis such as https://app.circleci.com/insights/github/emscripten-core/emscripten/workflows/build-test/tests?branch=xmlrunner

There are some caveats, such as that I don't know how it handles things such as differences across platforms or test variants. But it seems useful for a start.

File: .circleci/config.yml [hunk 1]
```diff
@@ -174,6 +174,8 @@ commands:
           command: |
             ./test/runner << parameters.test_targets >>
             $EMSDK_PYTHON ./tools/check_clean.py
+      - store_test_results:
+          path: test-results.xml
```
Comment (sbc100) [inline .circleci/config.yml:178]: How about `out/test-results.xml`..  I've been thinking about designating and using `out/` as place to put test artifacts.

File: .gitignore [hunk 2]
```diff
@@ -26,3 +26,6 @@ htmlcov/
 coverage.xml
 
 .DS_Store
+
+# CircleCI test result file
+test-results.xml
```
Comment (sbc100) [inline .gitignore:31]: ./out/

File: requirements-dev.txt [hunk 3]
```diff
@@ -22,3 +22,5 @@ docutils==0.17.1
 
 # Needed by test/test_sockets.py
 websockify==0.10.0
+
+unittest-xml-reporting
```
Comment (sbc100) [inline requirements-dev.txt:26]: Put this up the top in the first block?  and pin a certain version?

File: test/runner.py [hunk 4]
```diff
@@ -294,7 +294,17 @@ def run_tests(options, suites):
...
+
+  try:
+    import xmlrunner # type: ignore
+    # Output fd must be left open until after testRunner.run() below
+    output = open('test-results.xml', 'wb')
+    testRunner = xmlrunner.XMLTestRunner(output=output, verbosity=2,
+                                         failfast=options.failfast)
+    print('Writing XML test output to ' + os.path.abspath(output.name))
+  except ImportError:
+    testRunner = unittest.TextTestRunner(verbosity=2, failfast=options.failfast)
```
Comment (sbc100) [inline test/runner.py:306]: Its generally accepted that you have to run `pip install -r requirements-dev.txt` before you can run any tests.. so I don't think this needs to be conditional.

**Comment by dschuff** [comment]
@RReverser experimented with the same thing in https://github.com/emscripten-core/emscripten/commit/36d61b9eff458bd142c93ef408632e972ca06cb7

@RReverser is the fake traceback part related to the unittest part?
He did 2 things I didn't do in my version:
1) conditioned the XML bits on being in CI (based on checking the environment). I liked having the XML output locally too, but that was mostly because I wanted to look at the XML output :). Do we care about leaving it around when running locally? I figured we mostly wouldn't have xmlrunner installed locally anyway (I didn't).
2) gave the upload step an explicit name. Looking at my link to the test results above, the default test name seems to be "uploading test results" which seems clear enough to me.

**Comment by RReverser** [comment]
> @RReverser is the fake traceback part related to the unittest part?

Yeah without it XML runner was complaining about missing properties and I found tblib to have a more feature-complete version of FakeTraceback implementation (at least it fixed that issue for me).

> gave the upload step an explicit name

I actually did that to reuse the test collection across many workflows, basically wherever `prepare-for-tests` is used too.

In particular, you're adding it only to `run-tests` but browser tests don't use that step - they do something custom instead.

**Comment by kripken** [comment]
Sounds good. Should we be able to see the improved output here somehow (in links to the CI runs)? Or do we need those special links you provided before @dschuff ? (those do look useful)

**Comment by dschuff** [comment]
> Sounds good. Should we be able to see the improved output here somehow (in links to the CI runs)? Or do we need those special links you provided before @dschuff ? (those do look useful)

Yes, if you look at the build of 6161c9a (in e.g. test-core0), it has a test tab next to the 'steps' tab, which just lists the one failing test.

**Comment by dschuff** [comment]
There are also 'insights' across multiple runs, e.g. https://app.circleci.com/insights/github/emscripten-core/emscripten/workflows/build-test/tests?branch=xmlrunner

**Comment by kripken** [comment]
I see, thanks, that does lead to [a test-specific error](https://app.circleci.com/pipelines/github/emscripten-core/emscripten/24943/workflows/c59862aa-233c-4bd0-8b3c-2568cc71e37f/jobs/588349). It might be a little too specific though, as it's just the python stack trace - any logging from the test when it run isn't shown? But I guess we can flip back to the "steps" tab to see the entire output if that's necessary.

**Comment by RReverser** [comment]
> any logging from the test when it ran isn't shown?

For that you want https://github.com/emscripten-core/emscripten/issues/18308.

**Comment by dschuff** [comment]
> I see, thanks, that does lead to [a test-specific error](https://app.circleci.com/pipelines/github/emscripten-core/emscripten/24943/workflows/c59862aa-233c-4bd0-8b3c-2568cc71e37f/jobs/588349). It might be a little too specific though, as it's just the python stack trace - any logging from the test when it ran isn't shown? But I guess we can flip back to the "steps" tab to see the entire output if that's necessary.

Yes, you can always go back to the steps tab, which is the same as it always was. I agree that it would be nice to fix as @RReverser said. This is a minimal enough change on its own that I think it's still a nice stepping stone and worth trying out, especially to see if we find the test reporting useful.

**Comment by dschuff** [comment]
OK, I think this is working; any more reviews?

**Review by sbc100 — APPROVED** [review]
Does this change the output run I run locally? 

Does it mean the I will always generate XML output when I run the test locally?  Is there any cost to that?

**Comment by dschuff** [comment]
It doesn't change the output on stdout/stderr. It does always generate XML. I haven't noticed any cost, but you'd be more likely to notice if there was a cost than I would. In his version, @RReverser conditioned XML generation on a `CI` env var. We could do that if you'd prefer.

**Comment by RReverser** [comment]
> Does it mean the I will always generate XML output when I run the test locally? Is there any cost to that?

From Derek's description above, it does. Personally, I'd still limit it to CI (when CI env variable is set) to reduce noise - I don't think local XMLs are very useful outside of debugging CI integration - but up to you.

**Comment by kripken** [comment]
Limiting to CI using an env var seems reasonable to me.

**Review by kripken — APPROVED** [review]

**Review by sbc100 — APPROVED** [review]

**Comment by kripken** [comment]
I get this error when I try to run tests locally now:
```
$ ./test/runner core2.test_random_device
Traceback (most recent call last):
  File "test/runner.py", line 31, in <module>
    import xmlrunner # type: ignore
ModuleNotFoundError: No module named 'xmlrunner'
```
Is that expected - is it not a module present by default in python?

**Comment by sbc100** [comment]
> I get this error when I try to run tests locally now:
> 
> ```
> $ ./test/runner core2.test_random_device
> Traceback (most recent call last):
>   File "test/runner.py", line 31, in <module>
>     import xmlrunner # type: ignore
> ModuleNotFoundError: No module named 'xmlrunner'
> ```
> 
> Is that expected - is it not a module present by default in python?

You need to run `pip install -r requirements-dev.txt`.  

I've been thinking about adding some easier way to ensure folks have run this, and/or remind them to.

**Pull request merged**